schema/caveschema.xsd

```diff
@@ -1016,28 +1016,47 @@
     </xs:complexType>
   </xs:element>
 
-  <!-- Simple Types (string validation) -->
-  <!-- TODO: better validation here so regex isn't overly complicated in read_xml (31) -->
+
+  <!-- file (xs:string)
+    No validation done
+  -->
   <xs:simpleType name="file">
     <xs:annotation>
       <xs:documentation>A filename type</xs:documentation>
     </xs:annotation>
     <xs:restriction base="xs:string" />
   </xs:simpleType>
+
+  <!-- vector (xs:string)
+    Match "([float], [float], [float],)"
+    Ignore whitespace
+    Positive or negative floats are accepted
+    Each [float] matches one of two cases:
+      Starts with a number, eg: 5 or 5. or 5.0
+      Is a decimal, eg: .5
+  -->
   <xs:simpleType name="vector">
     <xs:annotation>
       <xs:documentation>vector attrib</xs:documentation>
     </xs:annotation>
     <xs:restriction base="xs:string">
       <xs:whiteSpace value="collapse" />
+      <xs:pattern value="\s*\(\s*-?(\d+\.?(\d+)?|(\d+)?\.\d+)\s*,\s*-?(\d+\.?(\d+)?|(\d+)?\.\d+)\s*,\s*\s*-?(\d+\.?(\d+)?|(\d+)?\.\d+)\s*\)\s*" />
     </xs:restriction>
   </xs:simpleType>
+
+  <!-- color (xs:string)
+    Match "[int], [int], [int]"
+    Ignore whitespace
+    Only positive integers are accepted 
+  -->
   <xs:simpleType name="color">
     <xs:annotation>
       <xs:documentation>color attrib</xs:documentation>
     </xs:annotation>
     <xs:restriction base="xs:string">
       <xs:whiteSpace value="collapse" />
+      <xs:pattern value="\s*(\d+)\s*,\s*(\d+)\s*,\s*\s*(\d+)\s*" />
```

Comment:
Are there spaces between the comma and the int (`10     ,`)? Do you wanna optionally match trailing commas?

Reply:
I added the optional trailing comma match. I decided to add as many spaces as I could just to validate things. Some of the xml projects aren't linted at all, I figure this will save some headache

Interestingly the `\s*` at the beginning and end of the regex didn't matter.  I guess the whitespace in the xml gets trimmed?

Reply:
This has to do with the structure of your regex! Your pattern asks to find a match for your regex anywhere in the string! If you add a `^` as the first character of your regex it'll make sure the pattern match starts at the first character. If you add a `$` as the last character of your regex it'll make sure the last character of your match is the last character of the string. If you use both `^` and `$` it'll make sure your string exactly matches your regex without any extra characters before or after the match

Reply:
Ouu very interesting! I've definitely seem them before but didn't pick out their usage when paired together. Very cool